### PR TITLE
Fix issue with Transition and children props merging

### LINKED ISSUE #494: React pose: Transition props forwarding bug
## Describe the bug

Props resolvers get stale props.

Stale values are passed here - https://github.com/Popmotion/popmotion/blob/009e777b90100c08e9dae16564babae14682eb2f/packages/react-pose/src/components/Transition/children.ts#L88-L89

This happens because we mutate the children, so when the child enters it receives `propsForChild` as its own props and later when the same child exits its now own props (received previously from `propsForChild`) override current `propsForChild` thus resulting in stale values received. 

## How to reproduce
Write a simple exit/enter animation with resolved value. Enter the item, remove it - observe the resolver calls.

## Expected behaviour
Current values being passed to resolvers.

### PR DESCRIPTION
**Pull request by Andarist**
closes #501 (its new version of the other PR), closes #496, fixes #494

**Note** I have not tested this code on any live example.

### CODE REVIEW

File: packages/react-pose/src/components/PoseElement/index.tsx [hunk 1]
```diff
@@ -82,23 +82,18 @@ class PoseElement extends React.PureComponent<PoseElementInternalProps> {
...
     const { getInitialPoseFromParent, pose, _pose, initialPose } = this.props;
 
     if (initialPose) {
       return initialPose;
     } else {
-      // Feels like this could probably be simpler
       const parentPose = getInitialPoseFromParent && getInitialPoseFromParent();
-      const thisPose = Array.isArray(pose) ? pose : [pose];
-      const thisInternalPose = Array.isArray(_pose) ? _pose : [_pose];
+      const initialPoses = (Array.isArray(parentPose) ? parentPose : [parentPose])
```
Comment (InventingWithMonster) [inline packages/react-pose/src/components/PoseElement/index.tsx]: 🎉🎉

File: packages/react-pose/src/components/Transition/children.ts [hunk 2]
```diff
@@ -34,22 +36,35 @@ const handleTransition = (
     enterAfterExit,
     flipMove,
     onRest,
-    ...propsForChild
+    ...propsForChilds
```
Comment (InventingWithMonster) [inline packages/react-pose/src/components/Transition/children.ts]: Should really be `propsForChild` (as they get applied per child) or `propsForChildren` which is the correct pluralisation. Either's fine by me

File: packages/react-pose/src/components/Transition/children.ts [hunk 3]
```diff
@@ -34,22 +36,35 @@ const handleTransition = (
...
     finishedLeaving,
     hasInitialized,
+    indexedChildren: prevChildren,
     scheduleChildRemoval,
   }: State
 ) => {
   targetChildren = makeChildList(targetChildren);
 
-  const children: Array<ReactElement<any>> = [];
+  const nextState: Partial<State> = {
```
Comment (InventingWithMonster) [inline packages/react-pose/src/components/Transition/children.ts]: TIL

**Comment by Andarist** [comment]
I've added additional fix to this PR - please take a look at the last commit too. Issue demonstrated here - https://codesandbox.io/s/x3yk2jy3vo - notice that all children stay in the DOM after their exit animations finish.

**Comment by Andarist** [comment]
Friendly ping :)

**Comment by InventingWithMonster** [comment]
Ah yeah I'll take a look this morning thanks. Got a talk coming up and its
taking up most my time atm

**Comment by InventingWithMonster** [comment]
If you go to `packages/react-pose` and run `yarn build` do you get type errors?

**Comment by InventingWithMonster** [comment]
I'm going to add a step to circleCI that tests builds

**Comment by InventingWithMonster** [comment]
Ok if you merge master back in that'll add these checks to the branch

**Comment by Andarist** [comment]
I've rebased and I'm trying to fix the type errors, but my TS knowledge is really poor.

It's also weird that I'm getting type errors in the parts of the code that I haven't touched (I'm seeing several `Parameter 'paramName' implicitly has an 'any' type.` - they are quite easily fixable, but I have to annotate parts of the code that were not annotated previously, relaying on inference? not sure what TS can infer though).

Also most commonly from what I've seen things like `Props` (and maybe `State` but aint sure about this one) are typed as interfaces and you are using types for them, which are nominal, rather than structural from what I understand, therefore it's harder for me to type some code lines as TS is complaining about some things not being assignable to i.e. type State.

Anyway - gonna try to make it work, but would appreciate some insight into why things work how they work now (I'm especially curious about the whole implicit any thing, do you see those errors?)

**Comment by InventingWithMonster** [comment]
I only see one error at a time with Rollup which is a bit whack-a-mole. I'm going to have a deeper look into it tomorrow. I'm not bound to using `type` for `Props` and `State`, if `interface` works better in this instance then we should use that.

Inference in TS is patchy imo. I tend towards it because its cleaner, but for instance it doesn't seem to recognise that `filter(Boolean)` will *never* return `void`, which is annoying.

**Comment by Andarist** [comment]
It seems that I've managed to fix TS errors. Still unsure why I had problems at some places (mainly this implicit any thing) - I'll comment under changed lines, maybe you will know the reason behind the TS logic. If not - then screw it for now, need to finally read TS docs anyway, so maybe I'll find some answers there 😄 

Please review the last commit though to see if I havent done anything too stupid.

File: packages/react-pose/src/components/PoseElement/index.tsx [hunk 4]
```diff
@@ -63,11 +63,11 @@ class PoseElement extends React.PureComponent<PoseElementInternalProps> {
    * =============================================
    */
   private childrenHandlers: PoseContextProps = {
-    registerChild: props => {
+    registerChild: (props: ChildRegistration) => {
```
Comment (Andarist) [inline packages/react-pose/src/components/PoseElement/index.tsx]: places like this surprise me, not sure why TS complains about this on my machine, but not on yours (and not on CI), looks like some config mismatch, but I shouldnt be using any extra config over what's in the repo here

**Comment by InventingWithMonster** [comment]
LGTM thanks for your hard work on this one!

**Pull request merged**